lib/diagnostics_channel.js

```diff
@@ -0,0 +1,281 @@
+        done: false
+      });
+    }
+
+    if (this.done) {
+      return PromiseResolve({
+        done: true
+      });
+    }
+
+    return new Promise((resolve) => {
+      ArrayPrototypePush(this.waiting, resolve);
+    });
+  }
+}
+
+class SpanMessage {
+  constructor(action, id, data) {
+    this.action = action;
+    this.id = id;
+    this.data = data;
+  }
+
+  // Augment with type: span to identify over the wire through JSON
+  toJSON() {
+    return {
+      ...this,
+      type: 'span'
+    };
+  }
+}
+
+let spanId = 0;
+
+class InactiveSpan {
+  message() {}
+  annotate() {}
+  end() {}
+}
+
+class Span {
+  constructor(channel, data) {
+    this.channel = channel;
+    this.id = ++spanId;
+    this.ended = false;
+
+    this.message('start', data);
+  }
+
+  message(action, data) {
+    if (this.ended) return;
+
+    if (action === 'end') {
+      this.ended = true;
+    }
+
+    this.channel.publish(new SpanMessage(action, this.id, data));
+  }
+
+  annotate(data) {
+    this.message('annotation', data);
+  }
+
+  end(data) {
+    this.message('end', data);
+  }
+
+  static aggregate(channel, map = identity, onComplete) {
+    const spans = new Map();
+
+    if (typeof onComplete !== 'function') {
+      onComplete = map;
+      map = identity;
+    }
+
+    function onMessage(message) {
+      // A span message may be pre or post serialization so identify either
+      if (message instanceof SpanMessage || message.type === 'span') {
+        let messages = MapPrototypeGet(spans, message.id);
+        if (!messages) {
+          messages = [];
+          MapPrototypeSet(spans, message.id, messages);
+        }
+
+        ArrayPrototypePush(messages, map(message));
+
+        if (message.action === 'end') {
+          onComplete(messages);
+          MapPrototypeDelete(spans, message.id);
+        }
+      }
+    }
+
+    channel.subscribe(onMessage);
+
+    return () => channel.unsubscribe(onMessage);
+  }
+}
+
+// TODO(qard): should there be a C++ channel interface?
+class ActiveChannel {
+  [SymbolAsyncIterator]() {
+    return new AsyncIterableChannel(this);
+  }
+
+  subscribe(subscription) {
+    ArrayPrototypePush(this._subscribers, subscription);
+  }
+
+  unsubscribe(subscription) {
+    const index = ArrayPrototypeIndexOf(this._subscribers, subscription);
+    if (index >= 0) {
+      ArrayPrototypeSplice(this._subscribers, index, 1);
+
+      // When there are no more active subscribers, restore to fast prototype.
+      if (!this._subscribers.length) {
+        // eslint-disable-next-line no-use-before-define
+        ObjectSetPrototypeOf(this, Channel.prototype);
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (let i = 0; i < this._subscribers.length; i++) {
+      try {
+        const onMessage = this._subscribers[i];
+        onMessage(data, this.name);
+      } catch (err) {
+        process.nextTick(() => {
+          triggerUncaughtException(err, false);
```

Comment:
why in nextTick()?

Reply:
So it doesn't interfere with the rest of the subscribers.